slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+    """
+    pages = client.conversations_replies(
+        channel=channel,
+        ts=thread_ts,
+        oldest=oldest,
+        inclusive=False,
+    )
+    for page in pages:
+        yield page["messages"]
+
+
+# Track users whose profile information has been updated in our DB and
+# doesn't need to be checked again.
+USERS_CACHE = {}
+
+
+def get_or_create_user(user_id):
+    try:
+        return USERS_CACHE[user_id]
+    except KeyError:
+        # Even if the user exists already in our db, they may have changed
+        # their information in slack so we need to check.
+        user_data = client.users_info(user=user_id)
+        obj, _ = SlackUser.objects.update_or_create(
+            id=user_id,
+            defaults={
+                "name": user_data.data["user"]["name"],
+                "real_name": user_data.data["user"].get("real_name", ""),
+                "email": user_data.data["user"]["profile"].get("email", ""),
+                "image_48": user_data["user"]["profile"].get("image_48", ""),
+            },
+        )
+        USERS_CACHE[user_id] = obj
+        return obj
+
+
+def should_track_message(message):
+    # These are not regular messages
+    # https://api.slack.com/events/message#subtypes
+    return message.get("subtype") in {None, "me_message"} and "bot_id" not in message
+
+
+def fill_channel_gap(gap, debug):
+    logger.info(
+        "Fetching channel history for %r (%r) in range (%s,%s)",
+        gap.channel.name,
+        gap.channel.id,
+        gap.oldest_message_ts,
+        gap.newest_message_ts,
+    )
+    pages = channel_messages_in_range(
+        channel=gap.channel.id,
+        latest=gap.newest_message_ts,
+        oldest=gap.oldest_message_ts,
+    )
+    first = True
+    for page in pages:
+        # use a separate transaction per page to allow restoring from an
+        # interrupted run.
+        with transaction.atomic():
+            for message in page:
+                if first and gap.newest_message_ts is None:
+                    gap.channel.last_update_ts = message["ts"]
+                    gap.channel.save()
+                    first = False
+                # Shrink the gap, but no need to save until we've finished this
+                # page (transactionally).
+                gap.newest_message_ts = message["ts"]
+
+                if not should_track_message(message):
+                    continue
+
+                if "user" in message:
+                    user = get_or_create_user(message["user"])
+                    if debug:
+                        gap.channel.seenmessage_set.create(ts=message["ts"])
+                    SlackActivityBucket.track_activity(gap.channel, user, message["ts"])
+
+                if message.get("thread_ts"):
+                    # Track this thread in the db to be able to check for
+                    # updates later.
+                    Thread.objects.create(
```

Comment:
Hm, I feel like I'm probably misunderstanding, but if there _are_ updates later, wouldn't we hit this and duplicate the thread object? Should this be an `update_or_create`?

Reply:
"updates" meaning new messages posted to the thread, which are not handled at all in this function. A thread can only be started once. It should definitely not be update_or_create because seeing the thread-starting message twice would be a logic error that the unique constraint violation error will alert us to.